### DISCUSSION TITLE
ESC keymap to go back to builtin

### DISCUSSION THREAD

**Discussion by SlartThunder** [ibhagwan/fzf-lua#1123] [Ideas]
Is there a way one can have a keymap to eg. ESC key that would navigate back to previous fzf-lua window?
The use case would be such that user:
1. User opens builtin window
2. User selects one of the options
3. On the new window open, if keymap (ESC) is clicked, it would go back to builtin

**Comment by ibhagwan** [comment]
You can add an action to global defaults that sets esc as builtin:
```lua
require("fzf-lua").setup({
  defaults  = {
    actions = { ["esc"] = function() require("fzf-lua").builtin() end }
  }
})
```

**Comment by SlartThunder** [comment]
I came up with the following code because there was a collateral issue with the code provided, that forbids the user to close fzf completely.

The following code works fine except for the fact that one looses the custom initialization set. Is there a way I could achieve that without loosing my initialization?

```lua
M.back_to_builtin = function()
  --defaults  = {
  --  actions = { ["esc"] = function() require("fzf-lua").builtin() end }
  --},

  if not M.comes_from_builtin then
    require("fzf-lua").setup({
	  defaults  = {
	    actions = { ["esc"] = M.back_to_builtin }
	  }
	})
  else
    require("fzf-lua").setup({
	  defaults  = {
	    actions = { ["esc"] = actions.dummy_abort }
	  }
	})
  end 
  
  M.comes_from_builtin = not M.comes_from_builtin
  require("fzf-lua").builtin()
end
```

**Comment by ibhagwan** [comment]
> I came up with the following code because there was a collateral issue with the code provided, that forbids the user to close fzf completely.

I didn’t have this issue as I’m using `<C-c>` to close, currently due to #1091 ctrl-c mapped to esc but it can be restored with a single auto command:
> https://github.com/ibhagwan/nvim-lua/blob/94751553b3b9e54301af21413137c6f1068fecf6/lua/autocmd.lua#L8-L17
```lua
    vim.api.nvim_create_autocmd("FileType", {
      pattern = "fzf",
      callback = function(e)
        vim.keymap.set("t", "<C-c>", "<C-c>", { buffer = e.buf, silent = true })
      end,
    })
```


> The following code works fine except for the fact that one looses the custom initialization set. Is there a way I could achieve that without loosing my initialization?

Change the signature to:
```lua
M.back_to_builtin = function(_, opts)
```

And call the function with passing the original opts:
```lua
require("fzf-lua").builtin(opts)
```

**Reply by SlartThunder** [reply]
I did not quite understand your reply, though I figure out a path by placing `true` as a second parameter when calling the `setup` function (based on your code)
Here you are, just in case it helps someone else!
Some understanding, I created an additional Lua file in nvim folder/lua/ called fzf-extension.lua, that I load after plugins. There I place the following code:
```lua
local M = {}
M.comes_from_builtin = false
M.back_to_builtin = function()
  --defaults  = {
  --  actions = { ["esc"] = function() require("fzf-lua").builtin() end }
  --},
  
  if not M.comes_from_builtin then
    require("fzf-lua").setup({
	  defaults  = {
	    actions = { ["esc"] = M.back_to_builtin }
	  }
	}, true)
  else
    require("fzf-lua").setup({
	  defaults  = {
	    actions = { ["esc"] = actions.dummy_abort }
	  }
	}, true)
  end 
  
  M.comes_from_builtin = not M.comes_from_builtin
  require("fzf-lua").builtin()
end
return M
```
Then, I can instantiate the extended feature function in my mappings lua file:
```lua
cmd('command! -nargs=0 -bang F :lua require\'fzf-extension\'.back_to_builtin()<cr>')
cmd ('cnoreabbrev f F')
```

Most probably it could be cleaner in another way…

Thanks so much @ibhagwan !! And for the plugin also, nice job!!

**Reply by ibhagwan** [reply]
I would do this a bit differently and avoid having to call setup each time to switch the action, this way you can avoid having a separate user command, I’ll write it for you when I get home.

**Comment by ibhagwan** [comment]
This would be a better soltution, add this to your original `setup` call and you don't have to call `setup` again or use your special user command and `cnoreabbrev`:
```lua
require("fzf-lua").setup({
  defaults = {
    actions = {
      ["esc"] = function(_, opts)
        -- builtin doesn't fill the `__INFO` variable
        if opts.__INFO then
          require("fzf-lua").builtin()
        end
      end,
    },
  },
  -- rest of your setup options
})
```

**Reply by SlartThunder** [reply]
mmmm… very interesting!
Though the behavior is not exactly the same:
* With this setup, when one is in a FZF window , pressing "esc" takes you to `FzfLua builtin`


(just clarifying: the `cnoreabbrev` I use it because my finger pressing Shift key is not always feeling like doing so, and then I get into different calls… In this manner If I typo `:F` for `:f` it will still launch FzfLua builtin)

**Reply by ibhagwan** [reply]
> With this setup, when one is in a FZF window , pressing "esc" takes you to FzfLua builtin

That's what I thought you wanted, did you want that just for builtin?

If so you can just do this for the builtin only, create a new user commad that runs this code, modifying the `default` action (of the builtin) and then calls the picker you want with an override for `esc`:
```lua
require("fzf-lua").builtin({
  actions = {
    ["default"] = function(sel)
      require("fzf-lua")[sel[1]]({
        actions = { ["esc"] = function() require("fzf-lua").builtin() end }
      })
    end,
  },
})
```

**Reply by SlartThunder** [reply]
Wonderful!! 
To tell you the truth, both use cases are worth…

**Reply by SlartThunder** [reply]
Sorry, @ibhagwan, I get back with this.
I've tried the latest snipet and it works very well, but when runnig `live_grep` I get the following error:
[Fzf-lua] fn_selected threw an error: 
```bash
.../.config/nvim/lua/plugins/fzf-lua.lua:62: attempt to call a nil value
stack traceback:
^.../.local/share/nvim/lazy/fzf-lua/lua/fzf-lua/core.lua:222: in function <....local/share/nvim/lazy/fzf-lua/lua/fzf-lua/core.lua:216>
^I~/.config/nvim/lua/plugins/fzf-lua.lua:62: in function 'action'
^I.../.local/share/nvim/lazy/fzf-lua/lua/fzf-lua/actions.lua:72: in function 'act'
^I.../.local/share/nvim/lazy/fzf-lua/lua/fzf-lua/core.lua:137: in function 'fn_selected'
^I.../.local/share/nvim/lazy/fzf-lua/lua/fzf-lua/core.lua:215: in function <....local/share/nvim/lazy/fzf-lua/lua/fzf-lua/core.lua:214>
^I[C]: in function 'xpcall'
^I.../.local/share/nvim/lazy/fzf-lua/lua/fzf-lua/core.lua:214: in function <...local/share/nvim/lazy/fzf-lua/lua/fzf-lua/core.lua:208>
```

**Reply by ibhagwan** [reply]
I can't reproduce this, the error seems to be coming from the snippet, the only thing I can think of is a non-existing command name, can you add `_G.dump("selected", sel)` as the first line of the action and then check `:messages` after a crash?

**Comment by SlartThunder** [comment]
I get the following message:
`"selected" { '   <name_of_the_file.yml:3:15:  group_key: "groups"' }`

Then followed with the previous stack trace I sent.
```bash
.../.config/nvim/lua/plugins/fzf-lua.lua:62: attempt to call a nil value
stack traceback:
^.../.local/share/nvim/lazy/fzf-lua/lua/fzf-lua/core.lua:222: in function <....local/share/nvim/lazy/fzf-lua/lua/fzf-lua/core.lua:216>
^I~/.config/nvim/lua/plugins/fzf-lua.lua:62: in function 'action'
^I.../.local/share/nvim/lazy/fzf-lua/lua/fzf-lua/actions.lua:72: in function 'act'
^I.../.local/share/nvim/lazy/fzf-lua/lua/fzf-lua/core.lua:137: in function 'fn_selected'
^I.../.local/share/nvim/lazy/fzf-lua/lua/fzf-lua/core.lua:215: in function <....local/share/nvim/lazy/fzf-lua/lua/fzf-lua/core.lua:214>
^I[C]: in function 'xpcall'
^I.../.local/share/nvim/lazy/fzf-lua/lua/fzf-lua/core.lua:214: in function <...local/share/nvim/lazy/fzf-lua/lua/fzf-lua/core.lua:208>
```

**Reply by ibhagwan** [reply]
It seems like you overwrote the `default` action in your config at `config/nvim/lua/plugins/fzf-lua.lua:62`

**Reply by SlartThunder** [reply]
Yes! I missunderstood your indication on your snipet and placed into the call to setup.
